### ISSUE TITLE
tidy-html5 does not understand MathML

### ISSUE THREAD

**Issue by fred-wang** [htacg/tidy-html5#130]
This is the counterpart of #70 for MathML. The issue was raised during the MathML integration into MediaWiki, so we had to use a "hack" to workaround this limitation. See https://gerrit.wikimedia.org/r/#/c/154083/.

@GerHobbelt did some work a long time ago to support SVG/MathML but that was never merged into master, see https://github.com/GerHobbelt/htmltidy and http://sourceforge.net/p/tidy/patches/84/.

Testcase:

      <!doctype html>
      <html>
        <head>
          <title>testcase</title>
          <meta charset="utf-8"/>
        </head>
        <body>
          <math><mfrac><msqrt><mi>x</mi></msqrt><mn>2</mn></mfrac></math>
        </body>
      </html>

Result: 

      line 8 column 11 - Error: <math> is not recognized!
      line 8 column 11 - Warning: discarding unexpected <math>
      line 8 column 17 - Error: <mfrac> is not recognized!
      line 8 column 17 - Warning: discarding unexpected <mfrac>
      line 8 column 24 - Error: <msqrt> is not recognized!
      line 8 column 24 - Warning: discarding unexpected <msqrt>
      line 8 column 31 - Error: <mi> is not recognized!
      line 8 column 31 - Warning: discarding unexpected <mi>
      line 8 column 36 - Warning: discarding unexpected </mi>
      line 8 column 41 - Warning: discarding unexpected </msqrt>
      line 8 column 49 - Error: <mn> is not recognized!
      line 8 column 49 - Warning: discarding unexpected <mn>
      line 8 column 54 - Warning: discarding unexpected </mn>
      line 8 column 59 - Warning: discarding unexpected </mfrac>
      line 8 column 67 - Warning: discarding unexpected </math>
      line 5 column 11 - Warning: <meta> proprietary attribute "charset"
      line 5 column 11 - Warning: <meta> lacks "content" attribute
      Info: Document content looks like HTML Proprietary
      12 warnings, 5 errors were found!

**Comment by geoffmcl** [comment]
Thanks for the report... Sorry for the delay in response...

As you may know the latest devlopment branch is develop-500
$ git checkout develop-500

But in checking the latest code, this 'bug' is still present, while after building a input test sample based on yours above the sample passes the validators... 

Have found w3.org some documentaions on MathML, but would seek more pointers... especially a neat list of tags (and attributes)... etc... everything helps...

Once I begin to understand MathML will work on adding support, and advise when done.

**Comment by geoffmcl** [comment]
Ok, have found https://github.com/GerHobbelt/htmltidy, and will check to see if it contains MathML support... ALWAYS helpful when others have done the implementation...

**Comment by fred-wang** [comment]
cc'ing @physikerwelt as he has done the work for MediaWiki.

You can check the MDN pages for documentation: https://developer.mozilla.org/en-US/docs/Web/MathML

MathML has two forms: "presentation" (http://www.w3.org/TR/MathML3/chapter3.html) and "content" (http://www.w3.org/TR/MathML3/chapter4.html). Note that the former is the most important for a browser context and is the one described on MDN.

The RelaxNG schema (http://www.w3.org/Math/RelaxNG/) provides a formal syntax for MathML and is used by the HTML5 validator (with some modifications to allow mixing HTML5 & MathML: https://github.com/validator/validator/tree/master/schema/mml3).

**Comment by physikerwelt** [comment]
I did not fix the problem as such. MediaWiki currently does not check the MathML at all, and trust that the MathML generated by Mathoid is correct. See https://gerrit.wikimedia.org/r/#/c/154083/15

**Comment by geoffmcl** [comment]
Hi Frédéric (fred-wang),

Re:  tidy-html5 does not understand SVG #70
and:  tidy-html5 does not understand MathML #130

Thanks for the links... I am sure these will help.

As mention found https://github.com/GerHobbelt/htmltidy,
which I cloned and built, and it certainly contains the
macros, enums, code and tables addressing these two issues.

He had also implemented his own support for html5 in general,
and some other features, so it is not a 5 minute cut and
paste job. Read no simple dev-5 patch...

I need to schedule a few days just for this... unless
someone else could pull develop-500 branch, and htmltidy
and effect the manual 'transfer' first ;=))

In other words would really appreciate any help on this
fairly simple, but long tedious task of extracting just
the svg and mathml from htmltidy, and re-coding it to
dev-5 form...

Regards,
Geoff.

**Comment by geoffmcl** [comment]
Thanks to Ger Hobbelt htmltidy fork, have now add initial MathML support.

And as a thanks to Ger each change for this support is marked with a [i_a]2 comment, like that used in his code.

As stated, this is an initial implementation, and needs more testing, and perhaps modifications, extension of the attribute table, etc, etc...

**Comment by fred-wang** [comment]
Thank you very much for this initial implementation. https://github.com/htacg/tidy-html5/issues/130#issue-41044973 seems to validate, now!

Could you say more about what is supported exactly... In the commit I only see the declaration for the math element, so I'm assuming tidy will accept arbitrary MathML content for now? Please let me know if you need help to set up the element/attribute tables (and where they are in the source).

I just gave a try with the torture tests (https://mdn.mozillademos.org/en-US/docs/Mozilla/MathML_Project/MathML_Torture_Test$samples/MathML_Torture_Test) and I get these errors:

1) `<math> proprietary attribute "display"` Strangely, this error does not happen for other attributes like "stretchy" so I suppose this has something to do with the fact that it concerns attributes on the root element.

2) `<mspace> attribute "width" has invalid value ...` Here I assume that width is already known by tidy for HTML elements, but the syntax is different for MathML.

3) `unescaped & or unknown entity` The official list of entities used by the HTML5 validator is http://www.w3.org/TR/xml-entity-names/#htmlmathml

**Comment by geoffmcl** [comment]
@fred-wang thanks for testing and reporting... this is GREAT...

As indicated this was a very basic implementation using code from Ger Hobbelt fork, and I TOO am still learning to understand his code ;=))

Rather than the usual ALL table driven TidyTag_XXXXX, TidyAttr_XXXX, etc he used an new mode he called 'namespace' parsing... with a new GetTokenMode OtherNamespace...

Yes, there is one, namely the TidyTag_MATH, and it has the 'usual' big defined line in tags.c 

tags.c(223):  { TidyTag_MATHML, "math", VERS_ELEM_MATHML, &TY_(W3CAttrsFor_MATHML)[0], (CM_INLINE|CM_BLOCK|CM_MIXED), TY_(ParseNamespace),NULL }, /* [i_a]2 */

So for attributes on it you can look at the W3CAttrsFor_MATHML, which is SMALL at present, and certainly has no TidyAttr_DISPLAY, which presently does not exist, so that needs to be defined and  added, along with maybe others, that may or may not already be defined...

Now the 'namespace' parsing is all handled in ParseNamespace, sort of handled as xml, and this is the service that I am trying to understand...

Basically it just marches forward, accumulating everything until it reaches `</math>`

Obviously at present it does NOT expect `<mspace width="thinmathspace">`, nor `<mo stretchy="false">`, etc, so has to be coded to include attribute collection...

Initially with no actual attribute validation, but there are generic services that can accumulate, parse attributes, including entities, and store them on the node for later output...

This would happen before the code, after gathering the attributes -
````
			/* add another child to the current parent */
			TY_(InsertNodeAtEnd)(parent, node);
````

The entities currently known to tidy are all stored in entities.c, and that list MAY be complete, or perhaps need to be extended...

As to the actual attribute validation and then value validation, maybe a new type of table needs to be designed for use by ParseNameSpace(), rather than the traditional mentioned above... or maybe we need to go the whole hog, and defined every MathML tag, together with an attribute table...

At present only some attributes have their own value parsers, like say -
attrs.c(98):  { TidyAttr_ALIGN, "align", CH_ALIGN }, /* varies by element */
which has its OWN value checker CH_ALIGN -> CheckAlign which ensures the value is one of a set -
ctmbstr const values[] = {"left", "right", "center", "justify", NULL};

Others have none, like CH_PCDATA -> NULL, which accepts EVERYTHING between the double quotes...

So decisions need to be made on each of these... how much should tidy validate... of course one day it is hoped 100% ;=)) but we must start small...

I will try to make some time to look deeper into this, but for sure would really appreciate any assistance with code, table, definitions, etc, etc...

This can be via a Pull Request from your clone, or even just a simple diff patch file...

Anyway, I hope this helps explain a little more...

**Comment by fred-wang** [comment]
Thanks, I'll take a look to this next Friday.

For entities.c, it seems that only those of HTML4 are defined, so I guess a VERSION_FROM50 should be added for the missing entity names. The list will become much bigger, so entitiesLookup should probably do a binary search.

Also, lang/rang use the HTML4 mapping, while modern browsers have all switched to the HTML5 one, so that should probably be fixed too (http://www.w3.org/TR/xml-entity-names/#diff-xhtml1).

**Comment by geoffmcl** [comment]
@fred-wang interesting we both looked at entities first ;=))

I downloaded htmlmathml.json and running it through a little perl script found some 2,125 compared to just 253 in entities.c, and that is without looking at the code point differences!

So what I see at this time is a new table, say static const entity entities5[] =, perhaps with some improved searching to be used initially ParseNameSpace(), but later for more...

On the other parts, during this week I will try to add the intial attribute parsing, probably initially without any validation...

Certainly appreciate your help on this.

**Comment by geoffmcl** [comment]
@fred-wang added the display attribute for the math element, but no value check yet...

Now working on not giving warnings for attributes on other math children... but need to think about this more...

Have done nothing about supporting the big mathml entities list...

**Comment by fred-wang** [comment]
So some analysis on entities:

1) General remark: MathML/SVG are generated from converters, editors etc so the most important point preventing tidy from rejecting valid HTML5 code... code formatting is less important. At least for Wikipedia's purpose only validation is important.

2) The only mapping change is rang/lang. Since modern browsers have moved to the HTML5 version, I think tidy should do that too.

3) Indeed looking a the number of lines in a file, there are 2125 entities in HTML5. A binary search will give a complexity of log(2125)/log(2) = 11 so it should be quite fast.

4) One complication is that they are two functions:

  ctmbstr TY_(EntityName)( uint charCode, uint versions );
  Bool    TY_(EntityInfo)( ctmbstr name, Bool isXml, uint* code, uint* versions );

  which would mean two tables (one ordered by char code, the other by entity name) to be able to perform a binary search. However, IIUC the first one is only used by pprint.c to produce entity names in the output source code. Per comment 1), I believe we don't care about that for now.

6) Another complication is that some entity name generates several characters, for example "vsupne" is "&#x0228B;&#x0FE00;". These entities will need more work since the callers should be adapted too...

For now, I'll just submit a pull request for 2) which is easy.

**Comment by geoffmcl** [comment]
@fred-wang thanks for the pull request...

1) Agree that as a start tidy should NOT reject anything in the math element. Have pushed a set of very kludgy fixes which I think does that.

2) Thanks for the mapping changes on rang/lang. This was merged as is.

3) I agree the bigger job is to now add a table with all 2125 html5 entities.

4) I would not initially go for speed. It still does not seem extreme to be doing a simple 2125 search. As to having two tables, maybe later the 2nd table could be generated from the first during init rather than hard coding two big table. But my idea is to just get it running initially, and consider speed up options later.

5) Yes, I know the fact that some have several characters means that the new table must go beyond just a single unit value, maybe having 2 uints for each, with one as zero, or something...

Anyway as stated, have pushed what I hope is a 'temporary' fix to get us through the initial stage. Tidy now seems to pass the 'torture_test' file without flagging any attribute or entity warnings...

Appreciate any further testing you can do, and any help getting this new entity table in place.

**Comment by fred-wang** [comment]
@geoffmcl @physikerwelt 

So I just gave a try with the develop-500 branch and https://en.wikipedia.org/wiki/Help:Displaying_a_formula (with MathML enabled in the preferences) and I don't see any MathML errors. The major problem seems to be a warning with the invalid use of the "meta" tag in the body, but that's an issue with MediaWiki (due to a bug in Chrome IIRC). Moritz, are you able to launch a test instance with tidy develop-500 on all MediaWiki math pages?

**Comment by physikerwelt** [comment]
I'm able to launch a test instance but I don't know how to switch the tidy version

**Comment by geoffmcl** [comment]
Hi @physikerwelt, not sure what OS you are in, but in linux, and OS X that is, in a terminal -

If you have already done 1. earlier, then you need to make sure you are in the develop-500 branch  
  $ cd tidy-html5
  $ git checkout develop-500
  $ git pull
  $ cd build/cmake

To start from scratch  
1. $ git clone git@github.com:htacg/tidy-html5.git # get the source
2. $ cd tidy-html5/build/cmake # get into the build directory
3. $ cmake ../.. # generate makfile
4. $ make # compile and link
6. $ [sudo] make install # optionally install

This should give you a working tidy5 binary, probably in /usr/local/bin... If you wanted it in /usr/bin then step 3. would be -  
3. $cmake ../.. -DCMAKE_INSTALL_PREFIX=/usr

In windows, step 4. can be -  
4. $ cmake --build . --config Release

Or you can use the make-gui... then MSVC IDE...

That should give you a tidy5.exe in the Release folder, which you could copy to any directory in your PATH environment variable to use globally...

But maybe I do not understand the problem...

**Comment by balthisar** [comment]
Will close. Support can open another issue if still not resolved.
